pkg/trace/sampler/probabilistic.go

```diff
@@ -42,21 +58,192 @@ type ProbabilisticSampler struct {
+		for k, v := range rule.Attributes {
+			compiledRules[i].attributes[k], err = regexp.Compile(v)
+			if err != nil {
+				return nil, fmt.Errorf("tag regex: key=%s, value=%s: %w", k, v, err)
+			}
+		}
+		compiledRules[i].scaledPercentage = uint32(rule.Percentage * percentageScaleFactor)
+		compiledRules[i].percentage = float64(rule.Percentage) / 100.
+	}
+	return compiledRules, nil
+}
+
+type probabilisticSamplerRules []probabilisticSamplerRule
+
+// String returns a string representation of the probabilisticSamplerRules.
+func (rs probabilisticSamplerRules) String() string {
+	rules := make([]string, len(rs))
+	for i, rule := range rs {
+		rules[i] = fmt.Sprintf("rule %d: %s", i, rule.String())
+	}
+	return strings.Join(rules, ", ")
+}
+
+type probabilisticSamplerRuleMetrics struct {
+	evaluations int64
+	matches     int64
+}
+
 // NewProbabilisticSampler returns a new ProbabilisticSampler that deterministically samples
 // a given percentage of incoming spans based on their trace ID
 func NewProbabilisticSampler(conf *config.AgentConfig) *ProbabilisticSampler {
 	hashSeedBytes := make([]byte, 4)
 	binary.LittleEndian.PutUint32(hashSeedBytes, conf.ProbabilisticSamplerHashSeed)
 	_, fullTraceIDMode := conf.Features["probabilistic_sampler_full_trace_id"]
+	rules, err := compileProbabilisticSamplerRules(conf.ProbabilisticSamplerTraceSamplingRules)
+	if err == nil && len(rules) > 0 {
+		log.Infof("Compiled probabilistic sampler trace sampling rules: %s", rules.String())
+	} else {
+		log.Errorf("Compiling probabilistic sampler trace sampling rules: %v", err)
+	}
 	return &ProbabilisticSampler{
 		enabled:                  conf.ProbabilisticSamplerEnabled,
 		hashSeed:                 hashSeedBytes,
 		scaledSamplingPercentage: uint32(conf.ProbabilisticSamplerSamplingPercentage * percentageScaleFactor),
 		samplingPercentage:       float64(conf.ProbabilisticSamplerSamplingPercentage) / 100.,
+		traceSamplingRules:       rules,
+		samplingRuleMetrics:      make(map[string]probabilisticSamplerRuleMetrics),
 		fullTraceIDMode:          fullTraceIDMode,
 	}
 }
 
+func (ps *ProbabilisticSampler) percentage(root *trace.Span) (uint32, float64) {
+	var matched, evaluated bool
+	defer func() {
+		if !evaluated {
+			return
+		}
+		ps.samplingRuleMetricsMutex.Lock()
+		metrics := ps.samplingRuleMetrics[root.Service]
+		metrics.evaluations++
+		if matched {
+			metrics.matches++
+		}
+		ps.samplingRuleMetrics[root.Service] = metrics
+		ps.samplingRuleMetricsMutex.Unlock()
+	}()
+	for _, rule := range ps.traceSamplingRules {
+		matched, evaluated = rule.evaluate(root)
+		if matched && evaluated {
+			return rule.scaledPercentage, rule.percentage
+		}
+	}
+	return ps.scaledSamplingPercentage, ps.samplingPercentage
+}
+
+func (ps *ProbabilisticSampler) report(statsd statsd.ClientInterface) {
+	if !ps.enabled || len(ps.traceSamplingRules) == 0 {
+		return
+	}
+	ps.samplingRuleMetricsMutex.Lock()
+	defer ps.samplingRuleMetricsMutex.Unlock()
```

Comment:
In order to release the mutex lock as soon as possible (and allow other goroutines to access it), consider copying the underlying metrics, reinitializing the `ps.samplingRuleMetrics`, and unlocking the mutex, then continue the processing and stats emission.